### ISSUE TITLE
graph: add Reversed method to Edge

### ISSUE THREAD

**Issue by kortschak** [gonum/gonum#825]
### Background
See  gonum/graph#87 (particularly https://github.com/gonum/graph/issues/87#issuecomment-113000044) and #824.

Currently when an undirected graph edge is returned there is a 0.5 chance that the from node of the edge is not the not the node that was used to query the graph. This can be resolved by the user when they know that this is possible and has code in place to handle this, but this can not be done for example in the case of `Copy` where the the input `src` and `dst` are essentially arbitrary graphs. In the case of `Copy` when an undirected graph is copied to a directed graph (if the `src` edge is to be used - see #824) then is it not possible to faithfully copy the edge according to the documentation (that a u--v becomes a pair of edges {u->v, v->u} which I would say is the only sensible approach).

If a graph were able to conditionally reverse an edge when needed, this problem goes away, and client code dealing with hops becomes simpler. This is only really needed for undirected graphs, but the operation makes sense as a general edge operation.

### Proposal
I propose that the `Edge` type be given an additional method `Reversed` that returns a copy of the receiver with the from and to nodes swapped, but all other characteristics remaining the same under the system the model is intended to be used in.

```
func (e edge) Reversed() graph.Edge {
    return edge{f: e.To(), t: e.From(), otherField: e.otherField}
}
```
By implication, the weighted edges and line types would also have this change.

The undirected interfaces would then be made to specify that edges returned are always returned such that the from and to node corresponds to the nodes used to query the `Graph.Edge` method (this is already the case for directed graphs).

### Potential impact of proposal

This is a breaking change; it requires an additional method to be added to all client code using edges or lines that are not provided by the gonum graph packages.
There also needs to be changes to the testgraph package to ensure that implementations satisfy the new semantic requirements.

The changes required in the gonum packages themselves are relatively minor and I would do them.

/cc @mewmew @vladimir-ch

**Comment by kortschak** [comment]
A working draft of this exists in https://github.com/gonum/gonum/compare/graph/reversed currently at 66a1d5964070f3ef8981265181ec81f2a0901b45. There is no documentation change or additional testing for the change yet.

One issue that arises is the need for `graph.Line` to have the `graph.Edge` return type for `Reversed`. I'm not sure if I'm upset about this; `Reversed` is not really for use outside the graph, so the naming clash is not very apparent, but it still looks ugly. It can be fixed with minor changes to testgraph, but those will conflict with #818, so I will not do that until that PR is merged.

**Comment by mewmew** [comment]
Without having given it too much thought, I must say that any proposal that adds more methods to one of the *core* graph interfaces, must really carry its weight. This would be similar to language additions in Go, where fixing subtleties or issues for a minority of use cases should not introduce a burden for the majority of the users.

Therefore, I would much rather explore other implementations of `Copy` that perhaps loop through the edges, and invoke `From` and `To` on those edges. Rather than to loop through the nodes. The suggestion in this last paragraph for how to change the implementation of `Copy` is just from the top of my head. I've not looked at the implementation of Copy in a while. However, I must say, I most strongly reserve myself from adding any new methods to the core graph interfaces, unless it has been a very commonly ocurring issue affecting most or all users. If not, we should find another day, even if say the implementation of `Copy` becomes less pretty.

**Comment by kortschak** [comment]
Note that the concern about orientation of edges in undirected graph querying is a long standing issue (I though I had skirted it, but it seems not). The reverse method is pretty close to a solution I had some time back, though I think this works better.

There is no widely available interface for obtaining edges from a graph (we do have one, but it does not live in graph), and if it did, it would still not help with the copy issue. It's not so much that it become less pretty, as that it becomes impossible without adding callbacks (which is a step too far for me in something that should just work).

**Comment by mewmew** [comment]
Not pretty, but passing the tests, also for undirected to directed:

```go
// Copy copies nodes and edges as undirected edges from the source to the destination
// without first clearing the destination. Copy will panic if a node ID in the source
// graph matches a node ID in the destination.
//
// If the source is undirected and the destination is directed both directions will
// be present in the destination after the copy is complete.
func Copy(dst Builder, src Graph) {
	nodes := src.Nodes()
	for nodes.Next() {
		dst.AddNode(nodes.Node())
	}
	nodes.Reset()
	for nodes.Next() {
		u := nodes.Node()
		to := src.From(u.ID())
		for to.Next() {
			v := to.Node()
			e := src.Edge(u.ID(), v.ID())
			dst.SetEdge(e)
			if _, ok := src.(Directed); !ok {
				// Source undirected.
				if _, ok := dst.(Directed); ok {
					// Destination directed.
					dst.SetEdge(dst.NewEdge(e.To(), e.From()))
				}
			}
		}
	}
}
```

```
u@x1 ~/g/s/g/v/g/graph> go test -v
=== RUN   TestCopy
--- PASS: TestCopy (0.00s)
=== RUN   TestCopyWeighted
--- PASS: TestCopyWeighted (0.00s)
=== RUN   TestUndirect
--- PASS: TestUndirect (0.00s)
=== RUN   TestUndirectWeighted
--- PASS: TestUndirectWeighted (0.00s)
PASS
ok  	gonum.org/v1/gonum/graph	0.006s
```

**Comment by kortschak** [comment]
That was something I considered, but it does not do the right thing for edges that are reversed. The value of the edge that's inserted contra is independent of the value in the `src` except for the endpoints. See discussion at #824.

BTW The condition is not necessary as it will get found on the pass through the nodes, but this does the same thing (still not enough)
```
func Copy(dst Builder, src Graph) {
	_, dstDirected := dst.(graph.Directed)
	nodes := src.Nodes()
	for nodes.Next() {
		dst.AddNode(nodes.Node())
	}
	nodes.Reset()
	for nodes.Next() {
		u := nodes.Node()
		to := src.From(u.ID())
		for to.Next() {
			v := to.Node()
			e := src.Edge(u.ID(), v.ID())
			if dstDirected || e.From().ID() == u.ID() {
				dst.SetEdge(e)
			} else {
				dst.SetEdge(dst.NewEdge(u, v))
			}
		}
	}
}
```

**Comment by kortschak** [comment]
An alternative is for undirected graphs to take an `UndirectedEdge interface { Edge; Reverse() Edge` when setting edges and store the edge as the reversible type. This would mean that only graph mutations would require this addition, but doubles the number of `SetEdge`-like methods.

**Comment by wsc0** [comment]
I have no strong opinions about this, but wanted to chime in with a couple of general conclusions I have come to w.r.t. graph libraries.

A looooong time ago I wrote the [stixar-graphlib](https://code.google.com/archive/p/stixar-graphlib/)

The way I handled this issue was below

```
public interface Edge extends Attributable
{
    /**
       Return the node from which this edge emanates.
       <b>Directed/Undirected Caveats:</b>
       <ul>
       <li>If this edge is a directed edge, this method will always return the same value.</li>
       <li>If this edge is an undirected edge, and is accessed through a (possibly empty) 
       sequence of calls to {@link #next} following a call to {@link Node#out} for a 
       node <tt>u</tt>, then this method will return <tt>u</tt>.
       </li>
       </ul>
     */
    public Node source();

    /**
       Return the node to which this edge points.
       <b>Directed/Undirected Caveats:</b>
       <ul>
       <li>If this edge is a directed edge, this method will always return the same value.</li>
       <li>If this edge is an undirected edge, and is accessed through a (possibly empty) 
       sequence of calls to {@link #next} following a call to {@link Node#out} for a 
       node <tt>u</tt>, then this method will return the vertex opposing <tt>u</tt>.
       </li>
       </ul>
     */
```

While pre-Go, I find the above Go-like in that the interface docs specify behaviour specific to certain contexts (directed/undirected).

The second conclusion was much more general.  I found that much of what I wrote in stixar-graphlib was over-general in the sense that I found it easier to _not_ use a library and just implicitly define "nodes" and "edges" based on what was there in existing data structures.  Then at least all the DFS/BFS is simpler than using a library (for me anyhoo).  Unfortunately, this did not extend to more complex algorithms like min/max flow.

Just wanted to add a little reference info to help inform your next steps.

**Comment by kortschak** [comment]
Closed by #867.
